Implement todo list and task (Ready)

LucasEduardoGiovanini opened the pull request:
this branch contains tests, use_cases package and the implementing about task and todo list.

LucasEduardoGiovanini committed:
when run, this project receives the implementations about complete a tas and undo a task. Also receives a basic structure of objection-orientation

LucasEduardoGiovanini committed:
implementing the to do list functions

LucasEduardoGiovanini committed:
Altering some packages and removing the prints of test

if run, this commit will change the addTask and removeTask functions to get a Task objetct like a parameter. Also i create a print function and i fix a bug.
In the task class, i fix the return of function getTask to convert the boolean stats for a string.

LucasEduardoGiovanini committed:
inserting a priority attribute in task and rename a todo list function

LucasEduardoGiovanini committed:
implementing some changes (made with Johnny)

if run, this commit rename the Task attribute "stats" to "completed' for make more sense, since the value of this attribute is boolean. Also change the function "getTask" to __repr__ function. Lastly, changes the print function of todolist to a __str__ function.

LucasEduardoGiovanini committed:
refactor the orderTasks function

if run, this commit refactor the orderTasks function to get the task that will be realocated and the task that will give the position for her.  the logic is: delete from the list tasks the task to be relocated, get index of task that will give the position and make a insert using the created index and the task.

LucasEduardoGiovanini committed:
partial creation of tests

If applied, this commit will make the partial creation of tests. Creating the lists and all necessary things to make the test independent.

LucasEduardoGiovanini committed:
Creating a use case package for call in the tests.

LucasEduardoGiovanini committed:
altering the use cases return

if run, this commit refactore some returns of functions to return the parameter.

LucasEduardoGiovanini committed:
altering some conditionals in use_case and finalizing the tests

i made a mistake making the conditional "if" validate by index if the value exists in the list, so i changed to see if the task is in the tasklists.

If applied, this commit will change what was explained above and call the use cases to tests.py to finalizing the tests functions.

LucasEduardoGiovanini committed:
removing unnecessary comments and fixing pep8 related issues

LucasEduardoGiovanini committed:
altering the use case conditional statement and calling in the test

LucasEduardoGiovanini committed:
Changing the test parameter and returning the created object in use case

johnnywell commented on domain/entities/todolist.py:
```diff
@@ -0,0 +1,24 @@
+        return "nome lista: " + self.name + " tarefas: " + str(self.tasks)
+
+
+    def orderTasks(self,task_to_be_relocated:Task,task_that_give_position:Task):
```
I think it's better api if you just receive the position instead of a task you want to replace. Also you need to handle the exception in case the list doesn't have that index.

johnnywell commented on domain/use_cases.py:
```diff
@@ -0,0 +1,42 @@
+from .entities.task import Task
```
Actually, I've say you should create only the entities. Maybe I wasn't clear, but use cases are not part of the domain.

johnnywell replied:
Any logic required for the entities to work must be in the entities methods you've create.  Most, if not, all the logic in this file must be part of you entities.

johnnywell replied:
Also, prefer handling exceptions than checking conditions, it's cheaper to ask forgiveness than permission.

johnnywell commented on tests/tests.py:
```diff
@@ -0,0 +1,69 @@
+import unittest
```
Your unit tests should test your entities, not use cases.

LucasEduardoGiovanini committed:
refactoring the orderTask function

if run, this commit will include a try catch statement for deal with errors about remove the task. The user can try delete a task that does not exist or try deleting pass a invalid index position.

LucasEduardoGiovanini committed:
Deleting the use_cases package

if run, this commit will delete use_cases.py because this package is no longer needed.

LucasEduardoGiovanini committed:
Creating a custom exception package.

If run, this commit will create a python package in the domain that contains a custom exception handling package. It will handle API errors.
Also he call this package in todolist.py for deal about an error of deleting a task that does not exist in a list.

LucasEduardoGiovanini committed:
refactoring some test functions and altering two return types in todolist

if run, this commit will refactor the tests functions to use my entities functions, now they deal with the errors.
making the refactorings, i realized that returns of addTask and removeTask functions its not making sense, so I returned a value that matches the operation done on the class.

LucasEduardoGiovanini committed:
refactoring some tests functions

if run, this commit will do the same as the past.

johnnywell commented:
Right, there's one thing still missing from your entities. But I think you'll spot it on the next task.
